Allow otherlibs not to be built

dra27 opened the pull request:
This is a precursor to allowing compilation of certain parts of the source tree without building the compiler.

The main addition is `--no-unix`, `--no-vmthreads` and `--no-str` options. Rather pleasingly, this is largely a configure script change - the other changes are a minor tweak in the build system to skip building caml-tex and changes to allow the testsuite to adapt to these libraries not being available.

One point which may want addressing further: we always build caml-tex, but we don't install it, which seems daft. Perhaps an option should be added so that it's tested on CI but not built by default?

dra27 committed:
Add hasunix ocamltest action

This test is skipped if *neither* unix nor win32unix have been built.

dra27 committed:
Add debugger ocamltest action

Skipped if the debugger has not been built.

dra27 committed:
Add hasstr and hassysthreads ocamltest actions

Skipped if the given library hasn't been built.

dra27 committed:
Update testsuite to work when libraries aren't built

dra27 committed:
Changes entry

shindere commented:
Thanks a lot, @dra27.

Say, do you wnat this in before the autoconf migration, or could it wait
a bit?

dra27 commented:
I would like it in 4.08, but I don't mind in which order!

The testsuite and ocamltest changes can of course be merged in advance of the change to configure itself, if you'd be happy to review those separately?

shindere commented:
Well giving it a seocnd thought I thought it would be more productive to
just review it right now so that it can be merged ASAP.

All the changes you propose look very good to me.

Thanks,in particular, for your work on making the library-related tests
mae sure the libraries they use have actually been built. It seeed I
forgot to do that, even for the debugger!

Regarding the names you used for configure arguments, I think those that
will actually be used by autoconf might differ a bit. In particular, I
think we will have `--without-something`rather than `--no-something` nso
I don't know whether you'd like to fix that now to avoid the names being
changed later, or whether you don't mind.

dra27 commented:
@shindere - thanks! I don't mind what the names of the parameters are. The current `configure` doesn't use GNU conventions, so I guess that the autoconf version is already going to be incompatible. Does that mean that you were already expecting existing ones like `--no-graph` to become `--without-graph` with no compatibility option? That seems fine to me.

Just for the full list, these are the options added:

 - `--with-graph`, `--with-unix`, `--with-str`, `--with-vmthreads`, `--with-systhreads` (all autoconf-compatible)
 - `--no-unix`, `--no-str`, `--no-vmthreads`, `--no-systhreads` (`--no-graph` already exists) would be `--without-*` in autoconf
 - `--debugger` would be `--with-debugger` (and the existing `--no-debugger` would become `--without-debugger`)
 - `--ocamldoc` would be `--with-ocamldoc` (and the existing `--no-ocamldoc` would be come `--without-ocamldoc`)

If the `--no-*` options are going to be removed and become `--without-*` I don't see any reason not to name all these ones here `--without-*` and cope with the inconsistency for the small amount of time before the existing ones get renamed too

shindere commented:
David Allsopp (2018/09/20 06:15 -0700):
> @shindere - thanks! I don't mind what the names of the parameters are.
> The current `configure` doesn't use GNU conventions, so I guess that
> the autoconf version is already going to be incompatible. Does that
> mean that you were already expecting existing ones like `--no-graph`
> to become `--without-graph` with no compatibility option?

Perhaps, yes, given that I am not sure yet how easy it would be to
provide backward-compatibility at this level.

> That seems fine to me.

Great.

> Just for the full list, these are the options added:
> 
>  - `--with-graph`, `--with-unix`, `--with-str`, `--with-vmthreads`,
>  - `--with-systhreads` (all autoconf-compatible)

Indeed.

>  - `--no-unix`, `--no-str`, `--no-vmthreads`, `--no-systhreads` (`--no-graph` already exists) would be `--without-*` in autoconf
>  - `--debugger` would be `--with-debugger` (and the existing `--no-debugger` would become `--without-debugger`)
>  - `--ocamldoc` would be `--with-ocamldoc` (and the existing `--no-ocamldoc` would be come `--without-ocamldoc`)
> 
> If the `--no-*` options are going to be removed and become
> `--without-*` I don't see any reason not to name all these ones here
> `--without-*` and cope with the inconsistency for the small amount of
> time before the existing ones get renamed too

You can either do that, or leave it as it is. I think it's really up to
you and what your personal preferences are.

I am going to aprove the PR so that you can feel free to merge it as
soon as you are happy. One thing I was wondering, though. Given that
there are several libraries, should we somehow reflect that we are
talking about libraries in the argument names? e.g. something like
--with-unix-libary, or --with-unix-lib? I am fine with things as they
are, just raising the question to make sure.

Octachron commented:
Concerning caml-tex, I agree that its status is a bit awkward. But since it depends on compiler-libs, it seemed better to raise an error as early as possible. Moving its build to a target build just before tests could be an option (after adding a bit of plumbing to the manual's Makefile). If I am not mistaken, it could be moved together with  `cmpbyt` which is also only used by the test suite.

shindere commented:
Florian Angeletti (2018/09/21 06:57 -0700):
> Concerning caml-tex, I agree that its status is a bit awkward. But
> since it depends on compiler-libs, it seemed better to raise an error
> as early as possible. Moving its build to a target build just before
> tests could be an option (after adding a bit of plumbing to the
> manual's Makefile). If I am not mistaken, it could be moved together
> with  `cmpbyt` which is also only used by the test suite.

Well cmpbyt is also used to make sure the bootstrap works. But perhaps
camltex could be moved to `testsuite/tools` ?

Octachron commented:
This sounds like a quite unnatural location for a manual's tool compared to either `tools` or `manual/tools`.

shindere commented:
Florian Angeletti (2018/09/21 11:37 -0700):
> This sounds like a quite unnatural location for a manual's tool
> compared to either `tools` or `manual/tools`.

Oh okay, I thought it would be easier to have it there but ifit's not
then no problem. :)

dra27 commented:
Indeed, it doesn’t seem to want moving, just that we don’t necessarily want it built unless it’s to be used (or for testing, of course)

dra27 commented:
@shindere - going back to the naming, it of course doesn't matter if autoconf gets in to 4.08 and this is merged with that, but perhaps it would be sensible for the naming in this GPR, being merged now, matches what the names would be in autoconf. So I'll change the new options added, as above, and add autoconf-compatible aliases for the ones which already exist (so `--without-ocamldoc`, `--without-debugger` and `--without-graph` added too).

dra27 commented:
Added as fixup commits for clarity of reviewing (I'll rebase before it's merged).

So I changed `--no` to `--without`. I like your idea with the `-lib` suffix too - so it's now the clearer `--with-graphics-lib`, etc. I left it as `--with-systhreads` and `--with-vmthreads` since these both define a library called threads and really that is morally more about enabling support for something than specifically a library.

One further question, which had been in the back of my mind, but I hadn't checked. We use things like `--with-debug-runtime` at the moment, but in autoconf world these would more normally be `--enable-debug-runtime` with `--with-*` used for external dependencies (for example, `--with-libunwind` or, as used to be, `--with-curses`).

What have you done for these so far? I should rename them to `--enable-*`/`--disable-*` if the plan would be to use [`AC_ARG_ENABLE`](https://www.gnu.org/software/autoconf/manual/autoconf-2.60/html_node/Package-Options.html#Package-Options) rather than `AC_ARG_WITH`.

shindere commented:
David Allsopp (2018/09/22 01:52 -0700):
> Added as fixup commits for clarity of reviewing (I'll rebase before
> it's merged).

Okay thanks!

> So I changed `--no` to `--without`. I like your idea with the `-lib`
> suffix too - so it's now the clearer `--with-graphics-lib`, etc. I
> left it as `--with-systhreads` and `--with-vmthreads` since these both
> define a library called threads and really that is morally more about
> enabling support for something than specifically a library.

Okay. I think this can be revisited anyway once the vmthreads have been
removed. Then there will be only one threads mechanism so we could keep
going with e.g. `--disable-threads`, meaning that they would be enabled
by default.

> One further question, which had been in the back of my mind, but I
> hadn't checked. We use things like `--with-debug-runtime` at the
> moment, but in autoconf world these would more normally be
> `--enable-debug-runtime` with `--with-*` used for external
> dependencies (for example, `--with-libunwind` or, as used to be,
> `--with-curses`).

> 
> What have you done for these so far?

Nothing, for the moment, but in a way I do tend to think it would be
good to stick to autoconf conventions for that. Not so much because they
are perfect, but rather because conventions help make decisions so why
not this one which is already in use and known by people that will have
to call `configure`.

> I should rename them to `--enable-*`/`--disable-*` if the plan would
> be to use
> [`AC_ARG_ENABLE`](https://www.gnu.org/software/autoconf/manual/autoconf-2.60/html_node/Package-Options.html#Package-Options)
> rather than `AC_ARG_WITH`.

Well if we stick to conventions strictly, that may result in us having
different argument styles for different libraries, I guess, depending on
whether they require an external package or not. So for instance
`graphics` would be a with-style, but since str does not depend on
anything external it should be an enable-style. Would that be okay or
would it be too inconsistent?

alainfrisch commented:
@shindere : since you approved this PR, is there anything blocking a merge (except for the conflicts)?

shindere commented:
Alain Frisch (2018/10/31 06:33 -0700):
> @shindere : since you approved this PR, is there anything blocking a
> merge (except for the conflicts)?

Well I don't know, but...

I'm about to submit the autoconf one, so I'd love not having to rebase,
if that's possible.

alainfrisch commented:
Ok, I thought you decided it was worth merging this one before yours.

shindere commented:
Alain Frisch (2018/10/31 13:45 +0000):
> Ok, I thought you decided it was worth merging this one before yours.

I did, but that was a few weeks ago now and things have changed a bit
meanwhile.

dra27 commented:
Rebased and updated. Sorry - this had slipped what with ICFP and all.

@shindere - I would like this to go into 4.08 if at all possible, though I'm very happy to have to rebase it onto the autoconf version

shindere commented:
David Allsopp (2018/10/31 14:21 -0700):
> Rebased and updated. Sorry - this had slipped what with ICFP and all.
> 
> @shindere - I would like this to go into 4.08 if at all possible,
> though I'm very happy to have to rebase it onto the autoconf version

@dra27 if you don't mind letting autoconf go in first I'd really
appreciate, yeah. I will be happily to help with the rebasing, because
the autoconf script is a bit messy at the moment.

damiendoligez commented:
@dra27 Do you have time to rebase this one this week?

shindere commented:
Damien Doligez (2019/01/07 06:26 -0800):
> @dra27 Do you have time to rebase this one this week?

I may try to do it if you (@dra27) so wish, just let me know.

dra27 committed:
Rename --enable-pthread to --enable-systhreads

Name better reflects the cross-platform nature of the library.

dra27 committed:
Make unix, vmthreads and str optional libraries

--disable-unix-lib, --disable-vmthreads and --disable-str-lib added to
prevent building these three libraries.

ocamldoc, the debugger and caml-tex are automatically disabled if their
prerequisites are not built. Using --enable-debugger and
--enable-ocamldoc will result in errors if these tools cannot be built.

dra27 committed:
Fix --disable-pthread

$systhread_support must be a valid OCaml boolean, as it gets written to
tools/config.ml

dra27 commented:
Sorry, only just back to work today! Rebased and updated - the ocamltest changes are unaltered, but the autoconf need reviewing. In particular, this starts the follow-on of some discussions in the original GPR for more appropriate names of some of the parameters (--enable-pthread renamed to --enable-systhreads, etc.)

nojb commented:
Sorry for coming in late to the discussion, but what about `--no-bigarray`?

dra27 commented:
@nojb - not at all, but do we really want that for the legacy library?

I tied it to --disable-unix-lib, as that's the actual dependency I think?

dra27 commented:
(so in other words `--disable-unix-lib` means unix isn't built **and** the legacy bigarray library - the module in the standard library is built regardless, obviously)

shindere commented:
Nicolás Ojeda Bär (2019/01/08 16:32 +0000):
> Sorry for coming in late to the discussion, but what about
> `--no-bigarray`?

AFAIK, autoconf does not support this namign scheme for arguments.

nojb commented:
My motivation is to have an easy way to **not** install the legacy `bigarray`; unless I am wrong one needs to edit the `Makefile` by hand to do this.

shindere commented:
Thanks, @dra27! Will review tomorrow afternoon at the latest.

dra27 committed:
Add --disable-bigarray-lib

Disables building the legacy bigarray library without also having to
disable the unix library.

dra27 commented:
@nojb - ah, I see. Extra `--disable-bigarray-lib` added

@shindere - thanks!

nojb commented:
Thanks!

shindere commented:
Just carefully went through the code and everything looks excellent to
me. Many thanks for such a nice piece of work. When merging, I'd suggest
to keep the commit history which is very clean. As far as I am
concerned, feel free to merge whenever you like. A particular thanks for
all the work done on the testsuite.

shindere commented:
Yeah, congrats!
